pkg/filesystem/subvolume.go

```diff
@@ -201,17 +208,98 @@ func unMarshaljson(list string) []fsStruct {
+		if err != nil {
+			logging.Fatal(err, "failed to delete subvolume of %q/%q/%q", fs, svg, subvol)
+		} else {
+			logging.Info("subvolume %q/%q/%q deleted", fs, svg, subvol)
+		}
 	} else {
-		logging.Info("subvolume %q is not stale", subvol)
+		logging.Info("subvolume %q/%q/%q is not stale", fs, svg, subvol)
+	}
+}
+
+func getMetadataPoolName(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespace, CephClusterNamespace, fs string) (string, error) {
+	// getFilesystem gets the filesystem
+	fsstruct, err := getFileSystem(ctx, clientsets, OperatorNamespace, CephClusterNamespace)
+	if err != nil {
+		return "", err
+	}
+
+	// this iterates over the filesystems and subvolumegroup to get the list of subvolumes that exist
+	for _, pool := range fsstruct {
+		if pool.Name == fs {
+			return pool.MetadataName, nil
+		}
+	}
+	return "", fmt.Errorf("pool not found")
+}
+
+// deleteOmap deletes omap object and key for the given subvolume.
+func deleteOmap(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespace, CephClusterNamespace, subVol, fs string) {
+	logging.Info("Deleting the omap object and key for subvolume %q", subVol)
+	omapdetails := getOmapDetails(ctx, clientsets, OperatorNamespace, CephClusterNamespace, subVol, fs)
+	poolName, err := getMetadataPoolName(ctx, clientsets, OperatorNamespace, CephClusterNamespace, fs)
+	if err != nil || poolName == "" {
+		logging.Fatal(fmt.Errorf("pool name not found: %q", err))
+	}
+	if omapdetails.omapval != "" {
+		// remove omap object.
+		_, err := exec.RunCommandInOperatorPod(ctx, clientsets, "rados", []string{"rm", omapdetails.omapval, "-p", poolName, "--namespace", "csi"}, OperatorNamespace, CephClusterNamespace, true)
+		if err != nil {
+			logging.Error(err, "failed to remove omap object for subvolume %q", subVol)
+		} else {
+			logging.Info("omap object:%q deleted", omapdetails.omapval)
+		}
+	}
+	if omapdetails.omapkey != "" {
+		// remove omap key.
+		_, err := exec.RunCommandInOperatorPod(ctx, clientsets, "rados", []string{"rmomapkey", "csi.volumes.default", omapdetails.omapkey, "-p", poolName, "--namespace", "csi"}, OperatorNamespace, CephClusterNamespace, true)
+		if err != nil {
+			logging.Error(err, "failed to remove omap key for subvolume %q", subVol)
+		} else {
+			logging.Info("omap key:%q deleted", omapdetails.omapkey)
+		}
 	}
 }
+
+// getOmapDetails gets the omap key and value details for a given subvolume.
+// Deletion of omap object required the subvolumeName which is of format
+// csi.volume.subvolume, where subvolume is the name of subvolume that needs to be
+// deleted.
+// similarly to delete of omap key requires csi.volume.ompakey, where
+// omapkey is the pv name which is extracted the omap object.
+func getOmapDetails(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespace, CephClusterNamespace, subVol, fs string) omapDetails {
```

Comment:
I'd suggest just returning the two strings. No need for a struct. Multiple return values is common in golang.
```suggestion
func getOmapDetails(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespace, CephClusterNamespace, subVol, fs string) (string, string) {
```